hedera-node/src/main/java/com/hedera/services/ledger/accounts/staking/RewardCalculator.java

```diff
@@ -20,66 +20,84 @@
 
 		if (stakePeriodStart > -1 && stakePeriodStart < todayNumber - 365) {
-			account.setStakePeriodStart(todayNumber - 365);
+			stakePeriodStart = todayNumber - 365;
 		}
 
-		stakePeriodStart = account.getStakePeriodStart();
-		if (isWithinRange(stakePeriodStart, todayNumber)) {
-			final long reward = computeReward(account, account.getStakedId(), todayNumber);
-			account.setStakePeriodStart(todayNumber - 1);
-			return reward;
+		if (isWithinRange(stakePeriodStart, todayNumber - 1)) {
+			final long reward = computeReward(account, account.getStakedId(), todayNumber, stakePeriodStart);
+			stakePeriodStart = todayNumber - 1;
+			this.accountReward = reward;
+		} else {
+			this.accountReward = 0L;
 		}
-		return 0;
+		this.accountUpdatedStakePeriodStart = stakePeriodStart;
 	}
 
-	boolean isWithinRange(final long stakePeriodStart, final long todayNumber) {
-		// if stakePeriodStart = -1 then it is not staked or staked to an account
-		// If it equals todayNumber, that means the staking changed today (later than the start of today),
-		// so it had no effect on consensus weights today, and should never be rewarded for helping consensus
-		// throughout today.  If it equals todayNumber-1, that means it either started yesterday or has already been
-		// rewarded for yesterday. Either way, it might be rewarded for today after today ends, but shouldn't yet be
-		// rewarded for today, because today hasn't finished yet.
-
-		return stakePeriodStart > -1 && stakePeriodStart < todayNumber - 1;
-	}
-
-	long computeReward(final MerkleAccount account, final long stakedNode, final long todayNumber) {
-		final var stakedNodeAccount = stakingInfo.get().get(EntityNum.fromLong(stakedNode));
+	long computeReward(final MerkleAccount account, final long stakedNode, final long todayNumber,
+			final long stakePeriodStart) {
+		final var stakedNodeAccount = stakeInfoManager.mutableStakeInfoFor(stakedNode);
 		final var rewardSumHistory = stakedNodeAccount.getRewardSumHistory();
-		final var stakePeriodStart = account.getStakePeriodStart();
+
 		// stakedNode.rewardSumHistory[0] is the reward for all days up to and including the full day todayNumber - 1,
 		// since today is not finished yet.
 		return account.isDeclinedReward() ? 0 :
 				account.getBalance() * (rewardSumHistory[0] - rewardSumHistory[(int) (todayNumber - 1 - (stakePeriodStart - 1))]);
 	}
+
+	public void updateRewardChanges(final MerkleAccount account, final Map<AccountProperty, Object> changes) {
+		computePendingRewards(account);
+
+		var balance = finalBalanceGiven(account, changes);
+
+		changes.put(BALANCE, balance + accountReward);
+		changes.put(STAKE_PERIOD_START, accountUpdatedStakePeriodStart);
+		rewardsPaid += accountReward; // used for adding balance change for 0.0.800
+	}
+
+	public void resetRewardsPaid() {
+		rewardsPaid = 0L;
+	}
+
+	public long rewardsPaidInThisTxn() {
+		return rewardsPaid;
+	}
+
+	public long getAccountReward() {
```

Comment:
These are just `@VisibleForTesting`?

Reply:
This is used in `StateView` for info queries. But `getAccountUpdatedStakePeriodStart` is only for tests. Added `@VisibleForTesting`

Reply:
Can be package-private now.